### PR TITLE
RyuJIT Tutorial

### PR DESCRIPTION
**Pull request by CarolEidt**
This is an initial translation of the slides presented at the PLDI 2016
RyuJIT Tutorial into markdown.
All of the diagrams are missing, and the IR dumps need to be updated to
reflect IR changes since the tutorial was given.

### CODE REVIEW

**Comment by CarolEidt** [comment]
I will probably add in the diagrams from the slides before merging. There's a lot more updating that should be done, but will probably be left for a future effort.
cc @dotnet/jit-contrib

**Comment by alpencolt** [comment]
cc @dotnet/arm32-contrib

File: Documentation/ryujit-tutorial.md [hunk 1]
```diff
@@ -0,0 +1,728 @@
...
+RyuJIT is the re-architected JIT for .NET.
+
+### Why “RyuJIT”?
+- Ryujin is a Japanese Sea Dragon
+
+We came up with the code name “RyuJIT” because our original code name had possible issues.
+We wanted something with “JIT” in the name, and the idea of a dragon came to mind because of the Dragon book that we all know and love.
+So – we just adapted the name of the Japanese sea dragon, Ryujin.
+
+### History
```
Comment (BruceForstall) [inline Documentation/ryujit-tutorial.md]: Specifically, JIT history (not RyuJIT)

File: Documentation/ryujit-tutorial.md [hunk 2]
```diff
@@ -0,0 +1,728 @@
...
+- Pre-dates .NET
+- Designed to support x86 only
+- Evolved, enhanced and extended over time & still evolving
+- Front-end: value numbering, SSA
```
Comment (BruceForstall) [inline Documentation/ryujit-tutorial.md]: These (front-end, back-end) aren't history

File: Documentation/ryujit-tutorial.md [hunk 3]
```diff
@@ -0,0 +1,728 @@
...
+#### Notes
+The JIT limits the number of lclVars for which it computes liveness. The limit is currently 512 for most targets, which is sufficient to include all of the lclVars for most methods. These are the only lclVars that will participate in optimization or register allocation.
+A lclVar whose address has been taken will never be tracked.
+They are also the lclVars for which GC ranges will be reported. Although the CLR has a precise GC model, the JIT can either report ranges over which a location (i.e. a register or stack location) contains a GC variable, or report a location as a GC value for the full method scope, and then ensure that its value either represents a valid heap location or null.
+
+### Liveness Analysis
+- The live-in and live-out sets are captured in the bbLiveIn and bbLiveOut fields of the BasicBlock.
+  - The GTF_VAR_DEF flag is set on a lclVar GenTree node that is a definition.
+  - The GTF_VAR_USEASG flag is set (in addition to the GTF_VAR_DEF flag) for the target of an update (e.g. +=).
+  - The GTF_VAR_USEDEF is set on the target of an assignment of a binary operator with the same lclVar as an operand.
```
Comment (BruceForstall) [inline Documentation/ryujit-tutorial.md]: `GTF_VAR_USEDEF` is gone

**Review by BruceForstall — APPROVED** [review]

File: Documentation/ryujit-tutorial.md [hunk 4]
```diff
@@ -0,0 +1,728 @@
...
+
+### .NET Runtime
+- An implementation of the Common Language Infrastructure [ECMA 335]
+  - Supports multiple languages, including C#, F# and VB
+- Sources are mostly shared between the “desktop” version and the open-source coreclr implementation:
+    http://www.github.com/dotnet/coreclr
+- RyuJIT is the “next generation” just in time compiler for .NET
+
+#### Notes
+For context, the .NET runtime has been around since about the turn of the millennium. It a virtual machine that supports the execution of a number of languages, primarily C#, Visual Basic, and F#.
```
Comment (mikedn) [inline Documentation/ryujit-tutorial.md]: It **is** a virtual machine...

File: Documentation/ryujit-tutorial.md [hunk 5]
```diff
@@ -0,0 +1,728 @@
...
+### Copy Propagation
+- Walks each block in the graph in dominator-first order
+  - When it encounters a variable with the same VN as a previously live variable, it is replaced with that variable
+
+### CSE & Assertion Propagation
+- Common Subexpression Elimination
+  - Redundant expressions are identified by value number
+  - If deemed profitable, they are evaluated to a new “temp” lclVar and then reused
+- Assertion Propagation
+  - Propagate properties such as constant values and non-nullness
```
Comment (mikedn) [inline Documentation/ryujit-tutorial.md]: Perhaps it should be mentioned by assertion propagation does eliminate some range checks as well

File: Documentation/ryujit-tutorial.md [hunk 6]
```diff
@@ -0,0 +1,728 @@
...
+- 
+### RyuJIT IR
+After ordering
+** Post-ordering IR diagram 
+
+#### Notes
+This is the same diagram as before, but with additional links to indicate execution order. The statements have a link to the first node in the list, and each node has a link to its previous and next node in the linear order.
+
+## RyuJIT Front-End
+
```
Comment (mikedn) [inline Documentation/ryujit-tutorial.md]: I don't see loop cloning and early propagation in this list of optimizations

File: Documentation/ryujit-tutorial.md (hunk 1 above)
Comment (4creators) [inline Documentation/ryujit-tutorial.md]: Etymology of project name is a very fine story - there are a lots of pictures with Ryujins which are quite similar to a dragon on Dragon Book cover :)

**Comment by jnm2** [comment]
This is stunningly cool to read. Thank you for writing this!
Probably the most newb question ever: I see the acronym IR used but not defined. What does it stand for?

**Comment by alpencolt** [comment]
@jnm2 IR - [Intermediate Representation](https://en.wikipedia.org/wiki/Intermediate_representation)

**Pull request merged**